### PR TITLE
Improve ODBC error handling on master

### PR DESCRIPTION
**Pull request by kovbal**
## What does this PR do?

This PR improves ODBC error handling, and provides a solution to the problem described in [this ticket](https://trac.osgeo.org/gdal/ticket/7258).

## What are related issues/pull requests?

These changes are backported to 1.11 in this PR: https://github.com/OSGeo/gdal/pull/499

## Tasklist

 - [ ] ADD YOUR TASKS HERE
 - [ ] Add test case(s)
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

## Environment

I tested it under:

* OS: Windows
* Compiler: MSVC 2017

### CODE REVIEW

File: gdal/port/cpl_odbc.h [hunk 1]
```diff
@@ -159,7 +159,7 @@ class CPLODBCStatement;
  */
 
 class CPL_DLL CPLODBCSession {
-    char      m_szLastError[SQL_MAX_MESSAGE_LENGTH + 1];
+    CPLString m_szLastError;
```
Comment (rouault) [inline gdal/port/cpl_odbc.h]: The variable should be renamed m_osLastError then

Comment (kovbal) [reply gdal/port/cpl_odbc.h]: You're right, will correct it.

File: gdal/port/cpl_odbc.cpp [hunk 2]
```diff
@@ -334,18 +333,30 @@ int CPLODBCSession::RollbackTransaction()
...
+        SQLSMALLINT nTextLength = 0;
+        SQLINTEGER nNativeError;
+
+        nDiagRetCode = SQLGetDiagRec( SQL_HANDLE_STMT, hStmt, nRecNum, achSQLState,
+                &nNativeError, reinterpret_cast<SQLCHAR *>(achCurErrMsg),
+                sizeof(achCurErrMsg)-1, &nTextLength );
+        if (nDiagRetCode == SQL_SUCCESS)
+        {
+            achCurErrMsg[nTextLength] = '\0';
+            m_szLastError += CPLString().Printf("%s[%5s]%s(%d)",
```
Comment (rouault) [inline gdal/port/cpl_odbc.cpp]:
```
cpl_odbc.cpp: In member function ‘int CPLODBCSession::Failed(int, HSTMT)’:
cpl_odbc.cpp:357:47: error: format ‘%d’ expects argument of type ‘int’, but argument 6 has type ‘SQLINTEGER {aka long int}’ [-Werror=format=]
                     achCurErrMsg, nNativeError);
                                               ^
```
So it looks the format string should probably be "%s[%5s]%s(%ld)"

**Comment by rouault** [comment]
@kovbal Do you intend to fix the above mentionned issues ?

**Comment by kovbal** [comment]
Yes, absolutely! Sorry for the long silence. Will correct the issues you
mentioned next week.

**Comment by rouault** [comment]
I've rebased and pushed your changes.

**Pull request closed without merging**